File: backend/src/OpenDriverHub.Infrastructure/Services/PartnerService.cs
using System.Text.Json;
using Microsoft.EntityFrameworkCore;
using OpenDriverHub.Application;
using OpenDriverHub.Domain;

namespace OpenDriverHub.Infrastructure.Services;

public class PartnerService : IPartnerService
{
    private readonly AppDbContext _db;
    public PartnerService(AppDbContext db) => _db = db;

    public async Task<List<ProductDto>> MyProductsAsync(Guid partnerId, CancellationToken ct)
        => (await _db.Products.Include(p => p.Partner)
                .Where(p => p.PartnerId == partnerId)
                .OrderBy(p => p.Title).ToListAsync(ct))
            .Select(p => p.ToDto()).ToList();

    public async Task<ProductDto> CreateProductAsync(Guid partnerId, ProductUpsertRequest req, CancellationToken ct)
    {
        var partner = await _db.Partners.FindAsync([partnerId], ct)
            ?? throw new AppException("Parceiro não encontrado.", 404);
        var p = new Product
        {
            PartnerId = partnerId,
            Title = req.Title,
            Description = req.Description,
            Price = req.Price,
            CashbackPercent = req.CashbackPercent,
            Kind = Mappings.ParseKind(req.Kind),
            ImageUrl = req.ImageUrl,
            Category = req.Category,
            Stock = req.Stock,
            Rating = 5.0,
            Partner = partner,
        };
        _db.Products.Add(p);
        await _db.SaveChangesAsync(ct);
        return p.ToDto();
    }

    public async Task<ProductDto> UpdateProductAsync(Guid partnerId, Guid productId, ProductUpsertRequest req, CancellationToken ct)
    {
        var p = await _db.Products.Include(x => x.Partner)
            .FirstOrDefaultAsync(x => x.Id == productId && x.PartnerId == partnerId, ct)
            ?? throw new AppException("Produto não encontrado.", 404);
        p.Title = req.Title;
        p.Description = req.Description;
        p.Price = req.Price;
        p.CashbackPercent = req.CashbackPercent;
        p.Kind = Mappings.ParseKind(req.Kind);
        p.ImageUrl = req.ImageUrl;
        p.Category = req.Category;
        p.Stock = req.Stock;
        await _db.SaveChangesAsync(ct);
        return p.ToDto();
    }

    public async Task DeleteProductAsync(Guid partnerId, Guid productId, CancellationToken ct)
    {
        var p = await _db.Products
            .FirstOrDefaultAsync(x => x.Id == productId && x.PartnerId == partnerId, ct)
            ?? throw new AppException("Produto não encontrado.", 404);
        p.Active = false; // soft-delete (preserva histórico de pedidos)
        await _db.SaveChangesAsync(ct);
    }

    public async Task<PartnerMetricsDto> MetricsAsync(Guid partnerId, CancellationToken ct)
    {
        var partner = await _db.Partners.FindAsync([partnerId], ct);
        var fee = partner?.FeePercent ?? 10m;

        // Itens deste parceiro (pedido multi-parceiro: agregamos por item).
        var items = await _db.OrderItems
            .Include(i => i.Order)
            .Where(i => i.PartnerId == partnerId)
            .ToListAsync(ct);

        bool IsValid(OrderItem i) =>
            i.Order!.Status is OrderStatus.Paid or OrderStatus.Redeemed;
        var valid = items.Where(IsValid).ToList();
        var revenue = valid.Sum(i => i.LineTotal);

        decimal NetOf(OrderItem i) => CommissionRules.PartnerNet(
            i.LineTotal, CommissionRules.PlatformFeeFor(i.LineTotal, fee),
            i.CashbackEarned);

        // Repasse REAL: "Recebido" = soma dos repasses lançados pelo admin.
        // "A receber" = líquido já resgatado ainda não repassado.
        var earnedNet = valid.Where(i => i.RedeemedAt != null).Sum(NetOf);
        var paidTransfer = await _db.PartnerPayouts
            .Where(p => p.PartnerId == partnerId)
            .SumAsync(p => (decimal?)p.Amount, ct) ?? 0m;
        var pendingTransfer = Math.Max(0m, earnedNet - paidTransfer);

        // Pedidos (distintos) que contêm itens deste parceiro.
        var orders = items.Select(i => i.Order!).DistinctBy(o => o.Id).ToList();
        var pendingCount = orders.Count(o => o.Status == OrderStatus.PendingPayment);
        var paidOrders = orders.Where(o => o.Status is OrderStatus.Paid or OrderStatus.Redeemed)
            .ToList();
        // "Resgatado" = pedido em que todos os itens deste parceiro foram resgatados.
        var redeemedCount = paidOrders.Count(o =>
            items.Where(i => i.OrderId == o.Id).All(i => i.RedeemedAt != null));
        var paidCount = paidOrders.Count - redeemedCount;

        var validOrderCount = paidOrders.Count;
        var avgTicket = validOrderCount > 0 ? Math.Round(revenue / validOrderCount, 2) : 0m;
        var cashbackGranted = valid.Sum(i => i.CashbackEarned);
        var uniqueCustomers = valid.Select(i => i.Order!.CustomerId).Distinct().Count();
        var redemptionRate = paidCount + redeemedCount > 0
            ? Math.Round(100m * redeemedCount / (paidCount + redeemedCount), 1)
            : 0m;

        var byHour = Enumerable.Range(0, 24)
            .Select(h => new SeriesPoint($"{h:00}h",
                valid.Count(i => i.Order!.CreatedAt.ToLocalTime().Hour == h)))
            .Where(s => s.Value > 0)
            .ToList();

        var last7 = Enumerable.Range(0, 7)
            .Select(i => DateTime.UtcNow.Date.AddDays(-6 + i))
            .Select(d => new SeriesPoint(
                d.ToString("dd/MM"),
                valid.Where(i => i.Order!.CreatedAt.Date == d).Sum(i => i.LineTotal)))
            .ToList();

        var topProducts = valid
            .GroupBy(i => i.ProductTitle.Length == 0 ? "—" : i.ProductTitle)
            .Select(g => new NamedValue(g.Key, g.Sum(i => i.LineTotal),
                g.Sum(i => i.Quantity)))
            .OrderByDescending(n => n.Value).Take(5).ToList();

        var byCategory = valid
            .GroupBy(i => i.Category.Length == 0 ? "—" : i.Category)
            .Select(g => new NamedValue(g.Key, g.Sum(i => i.LineTotal),
                g.Sum(i => i.Quantity)))
            .OrderByDescending(n => n.Value).ToList();

        var byMethod = valid
            .GroupBy(i => MethodLabel(i.Order!.PaymentMethod))
            .Select(g => new NamedValue(g.Key, g.Sum(i => i.LineTotal), g.Count()))
            .OrderByDescending(n => n.Value).ToList();

        return new PartnerMetricsDto(
            Math.Round(revenue, 2), validOrderCount,
            Math.Round(pendingTransfer, 2), Math.Round(paidTransfer, 2),
            avgTicket, Math.Round(cashbackGranted, 2), uniqueCustomers,
            pendingCount, paidCount, redeemedCount, redemptionRate,
            byHour, last7, topProducts, byCategory, byMethod);
    }

    private static string MethodLabel(PaymentMethod? m) => m switch
    {
        PaymentMethod.Pix => "Pix",
        PaymentMethod.CreditCard => "Crédito",
        PaymentMethod.DebitCard => "Débito",
        _ => "Outro",
    };

    public async Task<RedeemResult> RedeemAsync(Guid partnerId, Guid actorId, string code, bool confirm, CancellationToken ct)
    {
        var normalized = code.Replace("-", "").Trim().ToUpperInvariant();

        var order = await _db.Orders
                .Include(o => o.Customer)
                .Include(o => o.Items)
                .FirstOrDefaultAsync(o => o.Code.ToUpper() == normalized, ct)
            ?? throw new AppException("Código não encontrado.", 404);

        if (order.Status == OrderStatus.PendingPayment)
            throw new AppException("Voucher não está pago/liberado.", 409);
        if (order.Status == OrderStatus.Cancelled)
            throw new AppException("Pedido cancelado.", 409);

        // Só os itens DESTA loja (pedido pode ter vários parceiros).
        var myItems = order.Items.Where(i => i.PartnerId == partnerId).ToList();
        if (myItems.Count == 0)
            throw new AppException(
                "Este voucher não contém itens da sua loja.", 409);
        var pending = myItems.Where(i => i.RedeemedAt == null).ToList();
        if (pending.Count == 0)
            throw new AppException(
                "Os itens da sua loja neste voucher já foram resgatados.", 409);

        var partner = await _db.Partners.FindAsync([partnerId], ct)
            ?? throw new AppException("Parceiro não encontrado.", 404);
        var fee = partner.FeePercent;
        var subtotal = pending.Sum(i => i.LineTotal);
        var cashback = pending.Sum(i => i.CashbackEarned);
        var platformFee = CommissionRules.PlatformFeeFor(subtotal, fee);
        var net = CommissionRules.PartnerNet(subtotal, platformFee, cashback);
        var title = string.Join(", ",
            pending.Select(i => $"{i.Quantity}x {i.ProductTitle}"));

        if (!confirm)
            return new RedeemResult(order.Id, title, order.Customer!.Name,
                subtotal, fee, platformFee, cashback, net, false);

        var strategy = _db.Database.CreateExecutionStrategy();
        await strategy.ExecuteAsync(async () =>
        {
            await using var tx = await _db.Database.BeginTransactionAsync(ct);

            var fresh = await _db.Orders
                .Include(o => o.Items)
                .FirstOrDefaultAsync(o => o.Id == order.Id, ct)
                ?? throw new AppException("Pedido não encontrado.", 404);

            var freshPending = fresh.Items
                .Where(i => i.PartnerId == partnerId && i.RedeemedAt == null)
                .ToList();
            if (freshPending.Count == 0)
                throw new AppException(
                    "Os itens da sua loja neste voucher já foram resgatados.", 409);

            var now = DateTime.UtcNow;
            foreach (var item in freshPending)
            {
                item.RedeemedAt = now;
                var product = await _db.Products.FindAsync([item.ProductId], ct);
                if (product is not null && product.Stock >= item.Quantity)
                    product.Stock -= item.Quantity;
            }

            // Pedido só fica "Resgatado" quando TODOS os itens forem resgatados.
            if (fresh.Items.All(i => i.RedeemedAt != null))
            {
                fresh.Status = OrderStatus.Redeemed;
                fresh.RedeemedAt = now;
            }

            _db.Notifications.Add(new Notification
            {
                UserId = fresh.CustomerId,
                Title = "Voucher resgatado",
                Message =
                    $"Itens de {partner.Name} no pedido {fresh.Code} resgatados.",
            });
            _db.AuditLogs.Add(new AuditLog
            {
                ActorId = actorId,
                Action = "order.redeem",
                EntityType = "Order",
                EntityId = fresh.Id.ToString(),
                PayloadJson = JsonSerializer.Serialize(
                    new { fresh.Code, partnerId, net, cashback, subtotal }),
            });

            await _db.SaveChangesAsync(ct);
            await tx.CommitAsync(ct);
        });

        return new RedeemResult(order.Id, title, order.Customer!.Name,
            subtotal, fee, platformFee, cashback, net, true);
    }
}

File: backend/src/services/partnerService.ts
import type { NamedValue, SeriesPoint } from '../dtos/common.dto.js';
import type { ProductDto, ProductUpsertRequest, UpdateMyPartnerProfileRequest } from '../dtos/catalog.dto.js';
import type { AffiliatePartnerDto } from '../dtos/affiliate.dto.js';
import type { PartnerMetricsDto } from '../dtos/partner.dto.js';
import type { RedeemResult } from '../dtos/orders.dto.js';
import { driverCommissionFor, partnerNet, platformFeeFor, round2 } from '../domain/commissionRules.js';
import { AppError } from '../errors.js';
import { prisma } from '../infra/prisma.js';
import { parseProductKind, toAffiliatePartnerDto, toProductDto } from '../mappings.js';
import * as driverAffiliateService from './driverAffiliateService.js';

/** Autoatendimento: o próprio parceiro (loja ou afiliado) edita seus dados.
 * feePercent/active/asaasWalletId ficam de fora de propósito — exclusivos do
 * Admin (backend/src/services/adminService.ts updatePartner). */
export async function updateMyProfile(partnerId: string, req: UpdateMyPartnerProfileRequest): Promise<AffiliatePartnerDto> {
  const existing = await prisma.partner.findUnique({ where: { id: partnerId } });
  if (!existing) throw new AppError('Parceiro não encontrado.', 404);
  const updated = await prisma.partner.update({
    where: { id: partnerId },
    data: {
      ...(req.name != null ? { name: req.name } : {}),
      ...(req.segment != null ? { segment: req.segment } : {}),
      ...(req.logoUrl != null ? { logoUrl: req.logoUrl } : {}),
      ...(req.cnpj != null ? { cnpj: req.cnpj.trim() } : {}),
      ...(req.documentType != null ? { documentType: req.documentType } : {}),
      ...(req.city != null ? { city: req.city.trim() } : {}),
      ...(req.state != null ? { state: req.state.trim() } : {}),
      ...(req.lat != null ? { lat: req.lat } : {}),
      ...(req.lng != null ? { lng: req.lng } : {}),
    },
  });
  return toAffiliatePartnerDto(updated);
}

export async function myProducts(partnerId: string): Promise<ProductDto[]> {
  const rows = await prisma.product.findMany({
    where: { partnerId },
    include: { partner: true },
    orderBy: { title: 'asc' },
  });
  return rows.map((p) => toProductDto(p));
}

export async function createProduct(partnerId: string, req: ProductUpsertRequest): Promise<ProductDto> {
  const partner = await prisma.partner.findUnique({ where: { id: partnerId } });
  if (!partner) throw new AppError('Parceiro não encontrado.', 404);
  const p = await prisma.product.create({
    data: {
      partnerId,
      title: req.title,
      description: req.description,
      price: req.price,
      cashbackPercent: req.cashbackPercent,
      kind: parseProductKind(req.kind),
      imageUrl: req.imageUrl,
      category: req.category,
      stock: req.stock,
      rating: 5.0,
    },
    include: { partner: true },
  });
  return toProductDto(p);
}

export async function updateProduct(partnerId: string, productId: string, req: ProductUpsertRequest): Promise<ProductDto> {
  const existing = await prisma.product.findFirst({ where: { id: productId, partnerId } });
  if (!existing) throw new AppError('Produto não encontrado.', 404);
  const p = await prisma.product.update({
    where: { id: productId },
    data: {
      title: req.title,
      description: req.description,
      price: req.price,
      cashbackPercent: req.cashbackPercent,
      kind: parseProductKind(req.kind),
      imageUrl: req.imageUrl,
      category: req.category,
      stock: req.stock,
    },
    include: { partner: true },
  });
  return toProductDto(p);
}

export async function deleteProduct(partnerId: string, productId: string): Promise<void> {
  const existing = await prisma.product.findFirst({ where: { id: productId, partnerId } });
  if (!existing) throw new AppError('Produto não encontrado.', 404);
  // Soft-delete (preserva histórico de pedidos).
  await prisma.product.update({ where: { id: productId }, data: { active: false } });
}

function methodLabel(m: string | null): string {
  switch (m) {
    case 'Pix':
      return 'Pix';
    case 'CreditCard':
      return 'Crédito';
    case 'DebitCard':
      return 'Débito';
    default:
      return 'Outro';
  }
}

export async function metrics(partnerId: string): Promise<PartnerMetricsDto> {
  const partner = await prisma.partner.findUnique({ where: { id: partnerId } });
  const fee = partner?.feePercent.toNumber() ?? 10;

  // Itens deste parceiro (pedido multi-parceiro: agregamos por item).
  const items = await prisma.orderItem.findMany({ where: { partnerId }, include: { order: true } });

  const isValid = (i: (typeof items)[number]) => i.order.status === 'Paid' || i.order.status === 'Redeemed';
  const valid = items.filter(isValid);
  const revenue = valid.reduce((acc, i) => acc + i.lineTotal.toNumber(), 0);

  const netOf = (i: (typeof items)[number]) =>
    partnerNet(i.lineTotal.toNumber(), platformFeeFor(i.lineTotal.toNumber(), fee), i.cashbackEarned.toNumber());

  // Repasse REAL: "Recebido" = soma dos repasses lançados pelo admin.
  // "A receber" = líquido já resgatado ainda não repassado.
  const earnedNet = valid.filter((i) => i.redeemedAt !== null).reduce((acc, i) => acc + netOf(i), 0);
  const paidTransferAgg = await prisma.partnerPayout.aggregate({ where: { partnerId }, _sum: { amount: true } });
  const paidTransfer = paidTransferAgg._sum.amount?.toNumber() ?? 0;
  const pendingTransfer = Math.max(0, earnedNet - paidTransfer);

  // Pedidos (distintos) que contêm itens deste parceiro.
  const ordersById = new Map(items.map((i) => [i.order.id, i.order]));
  const orders = [...ordersById.values()];
  const pendingCount = orders.filter((o) => o.status === 'PendingPayment').length;
  const paidOrders = orders.filter((o) => o.status === 'Paid' || o.status === 'Redeemed');
  // "Resgatado" = pedido em que todos os itens deste parceiro foram resgatados.
  const redeemedCount = paidOrders.filter((o) => items.filter((i) => i.orderId === o.id).every((i) => i.redeemedAt !== null)).length;
  const paidCount = paidOrders.length - redeemedCount;

  const validOrderCount = paidOrders.length;
  const avgTicket = validOrderCount > 0 ? round2(revenue / validOrderCount) : 0;
  const cashbackGranted = valid.reduce((acc, i) => acc + i.cashbackEarned.toNumber(), 0);
  const uniqueCustomers = new Set(valid.map((i) => i.order.customerId)).size;
  const redemptionRate = paidCount + redeemedCount > 0 ? round2((100 * redeemedCount) / (paidCount + redeemedCount)) : 0;

  const byHour: SeriesPoint[] = Array.from({ length: 24 }, (_, h) => ({
    label: `${String(h).padStart(2, '0')}h`,
    value: valid.filter((i) => i.order.createdAt.getHours() === h).length,
  })).filter((s) => s.value > 0);

  const today = new Date();
  today.setUTCHours(0, 0, 0, 0);
  const last7: SeriesPoint[] = Array.from({ length: 7 }, (_, idx) => {
    const d = new Date(today);
    d.setUTCDate(d.getUTCDate() - 6 + idx);
    const next = new Date(d);
    next.setUTCDate(next.getUTCDate() + 1);
    const value = valid
      .filter((i) => i.order.createdAt >= d && i.order.createdAt < next)
      .reduce((acc, i) => acc + i.lineTotal.toNumber(), 0);
    return { label: `${String(d.getUTCDate()).padStart(2, '0')}/${String(d.getUTCMonth() + 1).padStart(2, '0')}`, value };
  });

  const topProducts = groupNamedValue(valid, (i) => (i.productTitle.length === 0 ? '—' : i.productTitle))
    .sort((a, b) => b.value - a.value)
    .slice(0, 5);
  const byCategory = groupNamedValue(valid, (i) => (i.category.length === 0 ? '—' : i.category)).sort(
    (a, b) => b.value - a.value,
  );
  const byMethod = groupNamedValueCount(valid, (i) => methodLabel(i.order.paymentMethod)).sort(
    (a, b) => b.value - a.value,
  );
  const driverReferral = await driverAffiliateService.storeMetrics(partnerId);

  return {
    totalRevenue: round2(revenue),
    totalSales: validOrderCount,
    pendingTransfer: round2(pendingTransfer),
    paidTransfer: round2(paidTransfer),
    averageTicket: avgTicket,
    cashbackGranted: round2(cashbackGranted),
    uniqueCustomers,
    pendingCount,
    paidCount,
    redeemedCount,
    redemptionRate,
    salesByHour: byHour,
    revenueLastDays: last7,
    topProducts,
    salesByCategory: byCategory,
    paymentMethods: byMethod,
    driverReferral,
  };
}

function groupNamedValue<T>(items: T[], keyOf: (i: T) => string): NamedValue[] {
  const map = new Map<string, { value: number; count: number }>();
  for (const item of items as any[]) {
    const key = keyOf(item);
    const entry = map.get(key) ?? { value: 0, count: 0 };
    entry.value += item.lineTotal.toNumber();
    entry.count += item.quantity;
    map.set(key, entry);
  }
  return [...map.entries()].map(([name, v]) => ({ name, value: round2(v.value), count: v.count }));
}

function groupNamedValueCount<T>(items: T[], keyOf: (i: T) => string): NamedValue[] {
  const map = new Map<string, { value: number; count: number }>();
  for (const item of items as any[]) {
    const key = keyOf(item);
    const entry = map.get(key) ?? { value: 0, count: 0 };
    entry.value += item.lineTotal.toNumber();
    entry.count += 1;
    map.set(key, entry);
  }
  return [...map.entries()].map(([name, v]) => ({ name, value: round2(v.value), count: v.count }));
}

export async function redeem(partnerId: string, actorId: string, code: string, confirm: boolean): Promise<RedeemResult> {
  const normalized = code.replace(/-/g, '').trim().toUpperCase();

  const order = await prisma.order.findFirst({
    where: { code: { equals: normalized, mode: 'insensitive' } },
    include: { customer: true, items: true },
  });
  if (!order) throw new AppError('Código não encontrado.', 404);

  if (order.status === 'PendingPayment') throw new AppError('Voucher não está pago/liberado.', 409);
  if (order.status === 'Cancelled') throw new AppError('Pedido cancelado.', 409);

  // Só os itens DESTA loja (pedido pode ter vários parceiros).
  const myItems = order.items.filter((i) => i.partnerId === partnerId);
  if (myItems.length === 0) throw new AppError('Este voucher não contém itens da sua loja.', 409);
  const pending = myItems.filter((i) => i.redeemedAt === null);
  if (pending.length === 0)
    throw new AppError('Os itens da sua loja neste voucher já foram resgatados.', 409);

  const partner = await prisma.partner.findUnique({ where: { id: partnerId } });
  if (!partner) throw new AppError('Parceiro não encontrado.', 404);
  const fee = partner.feePercent.toNumber();
  const subtotal = pending.reduce((acc, i) => acc + i.lineTotal.toNumber(), 0);
  const cashback = pending.reduce((acc, i) => acc + i.cashbackEarned.toNumber(), 0);
  const platformFee = platformFeeFor(subtotal, fee);
  // A comissão já foi de fato creditada ao motorista em
  // paymentService.approve() (não aqui) — isso só recalcula o mesmo valor
  // pra exibir o líquido certo pro parceiro nesta tela/no audit log.
  const commissionByPartner = await driverAffiliateService.commissionMapForOrder(order);
  const commissionInfo = commissionByPartner.get(partnerId);
  const commission = commissionInfo ? driverCommissionFor(subtotal, commissionInfo.percent) : 0;
  const net = partnerNet(subtotal, platformFee, cashback, commission);
  const title = pending.map((i) => `${i.quantity}x ${i.productTitle}`).join(', ');

  if (!confirm)
    return {
      orderId: order.id,
      productTitle: title,
      customerName: order.customer.name,
      paidPrice: round2(subtotal),
      feePercent: fee,
      platformFee: round2(platformFee),
      customerCashback: round2(cashback),
      partnerNet: round2(net),
      redeemed: false,
    };

  await prisma.$transaction(async (tx) => {
    const fresh = await tx.order.findUnique({ where: { id: order.id }, include: { items: true } });
    if (!fresh) throw new AppError('Pedido não encontrado.', 404);

    const freshPending = fresh.items.filter((i) => i.partnerId === partnerId && i.redeemedAt === null);
    if (freshPending.length === 0)
      throw new AppError('Os itens da sua loja neste voucher já foram resgatados.', 409);

    const now = new Date();
    const redeemedIds = new Set(fresh.items.filter((i) => i.redeemedAt !== null).map((i) => i.id));
    for (const item of freshPending) {
      await tx.orderItem.update({ where: { id: item.id }, data: { redeemedAt: now } });
      await tx.product.updateMany({
        where: { id: item.productId, stock: { gte: item.quantity } },
        data: { stock: { decrement: item.quantity } },
      });
      redeemedIds.add(item.id);
    }

    // Pedido só fica "Resgatado" quando TODOS os itens forem resgatados.
    const allRedeemed = redeemedIds.size === fresh.items.length;
    if (allRedeemed) {
      await tx.order.update({ where: { id: fresh.id }, data: { status: 'Redeemed', redeemedAt: now } });
    }

    await tx.notification.create({
      data: {
        userId: fresh.customerId,
        title: 'Voucher resgatado',
        message: `Itens de ${partner.name} no pedido ${fresh.code} resgatados.`,
      },
    });
    await tx.auditLog.create({
      data: {
        actorId,
        action: 'order.redeem',
        entityType: 'Order',
        entityId: fresh.id,
        payloadJson: JSON.stringify({ code: fresh.code, partnerId, net, cashback, subtotal, commission }),
      },
    });
  });

  return {
    orderId: order.id,
    productTitle: title,
    customerName: order.customer.name,
    paidPrice: round2(subtotal),
    feePercent: fee,
    platformFee: round2(platformFee),
    customerCashback: round2(cashback),
    partnerNet: round2(net),
    redeemed: true,
  };
}
